### ISSUE TITLE
IllegalStateException updating session inactive interval

### ISSUE THREAD

**Issue by araqnid** [eclipse/jetty.project#1284]
We upgraded an application that uses the Lift framework from Jetty 9.3.14 to 9.4.0. Since then we have had occasional occurrences of an IllegalStateException trying to lock the session within Session.setMaxInactiveInterval

We can't reproduce the error consistently, but it does correlate exactly with our 9.3->9.4 upgrade, which Istr changed a lot of the session implementation.

```
2017-01-25 08:56:30,953 ERROR [Jetty-388] n.l.h.LiftRules [Logging.scala:239] Exception being returned to browser when processing / java.lang.IllegalStateException: Locker is not reentrant
        at org.eclipse.jetty.util.thread.Locker.lock(Locker.java:47)
        at org.eclipse.jetty.server.session.Session.lock(Session.java:877)
        at org.eclipse.jetty.server.session.SessionHandler.sessionInactivityTimerExpired(SessionHandler.java:1302)
        at org.eclipse.jetty.server.session.Session$SessionInactivityTimeout.onIdleExpired(Session.java:127)
        at org.eclipse.jetty.io.IdleTimeout.checkIdleTimeout(IdleTimeout.java:166)
        at org.eclipse.jetty.io.IdleTimeout$1.run(IdleTimeout.java:50)
        at org.eclipse.jetty.io.IdleTimeout.activate(IdleTimeout.java:131)
        at org.eclipse.jetty.io.IdleTimeout.setIdleTimeout(IdleTimeout.java:102)
        at org.eclipse.jetty.server.session.Session$SessionInactivityTimeout.setIdleTimeout(Session.java:153)
        at org.eclipse.jetty.server.session.Session.setInactivityTimer(Session.java:547)
        at org.eclipse.jetty.server.session.Session.updateInactivityTimer(Session.java:525)
        at org.eclipse.jetty.server.session.Session.setMaxInactiveInterval(Session.java:478)
        at net.liftweb.http.provider.servlet.HTTPServletSession.setMaxInactiveInterval(HTTPServletSession.scala:37)
        at net.liftweb.http.LiftSession$$anonfun$fixSessionTime$1.apply(LiftSession.scala:853)
        at net.liftweb.http.LiftSession$$anonfun$fixSessionTime$1.apply(LiftSession.scala:844)
        at net.liftweb.common.Full.foreach(Box.scala:608)
        at net.liftweb.http.LiftSession.fixSessionTime(LiftSession.scala:844)
        at net.liftweb.http.LiftRules.net$liftweb$http$LiftRules$$_getLiftSession(LiftRules.scala:293)
        at net.liftweb.http.LiftRules$$anonfun$11.apply(LiftRules.scala:253)
        at net.liftweb.http.LiftRules$$anonfun$11.apply(LiftRules.scala:253)
        at net.liftweb.http.LiftServlet.getLiftSession(LiftServlet.scala:95)
        at net.liftweb.http.LiftServlet.doService(LiftServlet.scala:296)
        at net.liftweb.http.LiftServlet$$anonfun$doIt$1$1.apply(LiftServlet.scala:148)
        at net.liftweb.http.LiftServlet$$anonfun$doIt$1$1.apply(LiftServlet.scala:147)
        at net.liftweb.util.TimeHelpers$class.calcTime(TimeHelpers.scala:346)
        at net.liftweb.util.Helpers$.calcTime(Helpers.scala:34)
        at net.liftweb.util.TimeHelpers$class.logTime(TimeHelpers.scala:365)
        at net.liftweb.util.Helpers$.logTime(Helpers.scala:34)
        at net.liftweb.http.LiftServlet.doIt$1(LiftServlet.scala:147)
        at net.liftweb.http.LiftServlet.service(LiftServlet.scala:158)
        at net.liftweb.http.provider.HTTPProvider$$anonfun$service$2$$anonfun$apply$mcV$sp$1$$anonfun$apply$mcV$sp$2.apply$mcV$sp(HTTPProvider.scala:72)
        at net.liftweb.http.provider.HTTPProvider$$anonfun$service$2$$anonfun$apply$mcV$sp$1$$anonfun$apply$mcV$sp$2.apply(HTTPProvider.scala:71)
        at net.liftweb.http.provider.HTTPProvider$$anonfun$service$2$$anonfun$apply$mcV$sp$1$$anonfun$apply$mcV$sp$2.apply(HTTPProvider.scala:71)
        at net.liftweb.util.ThreadGlobal.doWith(ThreadGlobal.scala:71)
        at net.liftweb.http.URLRewriter$.doWith(Req.scala:1327)
        at net.liftweb.http.provider.HTTPProvider$$anonfun$service$2$$anonfun$apply$mcV$sp$1.apply$mcV$sp(HTTPProvider.scala:70)
        at net.liftweb.http.provider.HTTPProvider$$anonfun$service$2$$anonfun$apply$mcV$sp$1.apply(HTTPProvider.scala:70)
        at net.liftweb.http.provider.HTTPProvider$$anonfun$service$2$$anonfun$apply$mcV$sp$1.apply(HTTPProvider.scala:70)
        at net.liftweb.util.ThreadGlobal.doWith(ThreadGlobal.scala:71)
        at net.liftweb.http.provider.HTTPProvider$$anonfun$service$2.apply$mcV$sp(HTTPProvider.scala:66)
        at net.liftweb.http.provider.HTTPProvider$$anonfun$service$2.apply(HTTPProvider.scala:60)
        at net.liftweb.http.provider.HTTPProvider$$anonfun$service$2.apply(HTTPProvider.scala:60)
        at net.liftweb.util.ThreadGlobal.doWith(ThreadGlobal.scala:71)
        at net.liftweb.http.provider.HTTPProvider$class.service(HTTPProvider.scala:60)
        at net.liftweb.http.LiftFilter.service(LiftServlet.scala:928)
        at net.liftweb.http.provider.servlet.ServletFilterProvider$$anonfun$doFilter$1$$anonfun$apply$mcV$sp$1$$anonfun$apply$mcV$sp$2.apply$mcV$sp(ServletFilterProvider.scala:82)
        at net.liftweb.http.provider.servlet.ServletFilterProvider$$anonfun$doFilter$1$$anonfun$apply$mcV$sp$1$$anonfun$apply$mcV$sp$2.apply(ServletFilterProvider.scala:82)
        at net.liftweb.http.provider.servlet.ServletFilterProvider$$anonfun$doFilter$1$$anonfun$apply$mcV$sp$1$$anonfun$apply$mcV$sp$2.apply(ServletFilterProvider.scala:82)
        at net.liftweb.http.provider.servlet.ServletFilterProvider$class.handleLoan$1(ServletFilterProvider.scala:59)
        at net.liftweb.http.provider.servlet.ServletFilterProvider$class.net$liftweb$http$provider$servlet$ServletFilterProvider$$handleLoanWrappers(ServletFilterProvider.scala:63)
        at net.liftweb.http.provider.servlet.ServletFilterProvider$$anonfun$doFilter$1$$anonfun$apply$mcV$sp$1.apply$mcV$sp(ServletFilterProvider.scala:82)
        at net.liftweb.http.provider.servlet.ServletFilterProvider$$anonfun$doFilter$1$$anonfun$apply$mcV$sp$1.apply(ServletFilterProvider.scala:77)
        at net.liftweb.http.provider.servlet.ServletFilterProvider$$anonfun$doFilter$1$$anonfun$apply$mcV$sp$1.apply(ServletFilterProvider.scala:77)
        at net.liftweb.http.CoreRequestVarHandler$$anonfun$apply$16$$anonfun$apply$17$$anonfun$apply$18$$anonfun$apply$19.apply(Vars.scala:611)
        at net.liftweb.util.ThreadGlobal.doWith(ThreadGlobal.scala:71)
        at net.liftweb.http.CoreRequestVarHandler$$anonfun$apply$16$$anonfun$apply$17$$anonfun$apply$18.apply(Vars.scala:610)
        at net.liftweb.util.ThreadGlobal.doWith(ThreadGlobal.scala:71)
        at net.liftweb.http.CoreRequestVarHandler$$anonfun$apply$16$$anonfun$apply$17.apply(Vars.scala:609)
        at net.liftweb.util.ThreadGlobal.doWith(ThreadGlobal.scala:71)
        at net.liftweb.http.CoreRequestVarHandler$$anonfun$apply$16.apply(Vars.scala:608)
        at net.liftweb.util.ThreadGlobal.doWith(ThreadGlobal.scala:71)
        at net.liftweb.http.CoreRequestVarHandler$class.apply(Vars.scala:607)
        at net.liftweb.http.RequestVarHandler$.apply(Vars.scala:507)
        at net.liftweb.http.provider.servlet.ServletFilterProvider$$anonfun$doFilter$1.apply$mcV$sp(ServletFilterProvider.scala:75)
        at net.liftweb.http.provider.servlet.ServletFilterProvider$$anonfun$doFilter$1.apply(ServletFilterProvider.scala:75)
        at net.liftweb.http.provider.servlet.ServletFilterProvider$$anonfun$doFilter$1.apply(ServletFilterProvider.scala:75)
        at net.liftweb.http.CoreRequestVarHandler$$anonfun$apply$16$$anonfun$apply$17$$anonfun$apply$18$$anonfun$apply$19.apply(Vars.scala:611)
        at net.liftweb.util.ThreadGlobal.doWith(ThreadGlobal.scala:71)
        at net.liftweb.http.CoreRequestVarHandler$$anonfun$apply$16$$anonfun$apply$17$$anonfun$apply$18.apply(Vars.scala:610)
        at net.liftweb.util.ThreadGlobal.doWith(ThreadGlobal.scala:71)
        at net.liftweb.http.CoreRequestVarHandler$$anonfun$apply$16$$anonfun$apply$17.apply(Vars.scala:609)
        at net.liftweb.util.ThreadGlobal.doWith(ThreadGlobal.scala:71)
        at net.liftweb.http.CoreRequestVarHandler$$anonfun$apply$16.apply(Vars.scala:608)
        at net.liftweb.util.ThreadGlobal.doWith(ThreadGlobal.scala:71)
        at net.liftweb.http.CoreRequestVarHandler$class.apply(Vars.scala:607)
        at net.liftweb.http.TransientRequestVarHandler$.apply(Vars.scala:520)
        at net.liftweb.http.provider.servlet.ServletFilterProvider$class.doFilter(ServletFilterProvider.scala:74)
        at net.liftweb.http.LiftFilter.doFilter(LiftServlet.scala:928)
        at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1621)
        at org.jboss.resteasy.plugins.server.servlet.FilterDispatcher.doFilter(FilterDispatcher.java:66)
        at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1621)
        at org.eclipse.jetty.servlet.ServletHandler.doHandle(ServletHandler.java:541)
        at org.eclipse.jetty.server.handler.ScopedHandler.nextHandle(ScopedHandler.java:188)
        at org.eclipse.jetty.server.session.SessionHandler.doHandle(SessionHandler.java:1584)
        at org.eclipse.jetty.server.handler.ScopedHandler.nextHandle(ScopedHandler.java:188)
        at org.eclipse.jetty.server.handler.ContextHandler.doHandle(ContextHandler.java:1228)
        at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:168)
        at org.eclipse.jetty.servlet.ServletHandler.doScope(ServletHandler.java:481)
        at org.eclipse.jetty.server.session.SessionHandler.doScope(SessionHandler.java:1553)
        at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:166)
        at org.eclipse.jetty.server.handler.ContextHandler.doScope(ContextHandler.java:1130)
        at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:141)
        at org.eclipse.jetty.server.handler.StatisticsHandler.handle(StatisticsHandler.java:169)
        at org.eclipse.jetty.server.handler.gzip.GzipHandler.handle(GzipHandler.java:530)
        at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:132)
        at org.eclipse.jetty.server.Server.handle(Server.java:564)
        at org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:318)
        at org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:251)
        at org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:279)
        at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:112)
        at org.eclipse.jetty.io.ChannelEndPoint$2.run(ChannelEndPoint.java:124)
        at org.eclipse.jetty.util.thread.Invocable.invokePreferred(Invocable.java:122)
        at org.eclipse.jetty.util.thread.strategy.ExecutingExecutionStrategy.invoke(ExecutingExecutionStrategy.java:58)
        at org.eclipse.jetty.util.thread.strategy.ExecuteProduceConsume.produceConsume(ExecuteProduceConsume.java:201)
        at org.eclipse.jetty.util.thread.strategy.ExecuteProduceConsume.run(ExecuteProduceConsume.java:133)
        at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:672)
        at org.eclipse.jetty.util.thread.QueuedThreadPool$2.run(QueuedThreadPool.java:590)
        at java.lang.Thread.run(Thread.java:745)
```

**Comment by araqnid** [comment]
@janbartel I stepped into the IdleTimeout class and found that

setIdleTimeout is called initially with `old` 0 and `idleTimeout` -1 (when the session is being created)
then called once on the above code path with `old` -1 and `new` 120000 (2 minutes)
then called again on the above path with `old` 120000 and `new` 240000

(I'm not convinced Lift understands that HttpSession.getMaxInactiveInterval can return <= 0, but that doesn't seem to be producing ridiculous values at this time)

**Comment by janbartel** [comment]
@araqnid  hmmm, well we already have a test case where we create a session in one request, and then increase the maxInactiveInterval in a subsequent request, which is working fine. So there's some piece of info missing here. Can you describe the scenario where you update the maxInactiveInterval? For example, is it that you are creating a session and then modifying the maxInactiveInterval in the same request?  Is this a clustered session? What are your SessionCache settings?

**Comment by janbartel** [comment]
Also can you turn on debug level for org.eclipse.jetty.server.session please.

**Comment by araqnid** [comment]
This is the chunk of code from Lift:

```scala
  def fixSessionTime(): Unit = synchronized {
    for (httpSession <- this.httpSession) {
      lastServiceTime = millis // DO NOT REMOVE THIS LINE!!!!!
      val diff = lastServiceTime - httpSession.lastAccessedTime // -> HttpSession.getLastAccessedTime()
      val maxInactive = httpSession.maxInactiveInterval.toInt  // -> HttpSession.getMaxInactiveInterval()
      val togo: Int = maxInactive - (diff / 1000L).toInt
      // if we're within 2 minutes of session timeout and
      // the Servlet session doesn't seem to have been updated,
      // extends the lifespan of the HttpSession
      if (diff > 1000L && togo < 120) {
        httpSession.setMaxInactiveInterval(maxInactive + 120)
      }
    }
  }
```

we're not doing anything unusual for settings, just creating a default SessionHandler:

```scala
    servletContext.setSessionHandler(new SessionHandler())
```

I'll enable the debug logging and try again to reproduce the failure locally (so far my only luck with that was when I was stepping through with the debugger)

**Comment by araqnid** [comment]
I managed to reproduce the issue locally, just once so far. Still not sure what triggered it; as I remember, I went back to an test browser I hadn't used for a while, but according to the log this wasn't a particularly new session.

Looks like the session idle timer was being set for the first time on a session, although the session had been used as an immortal for some time.

[Pops.log.txt](https://github.com/eclipse/jetty.project/files/750743/Pops.log.txt)
